src/vmspawn/vmspawn-register.c

```diff
@@ -8,33 +9,39 @@
 #include "macro.h"
 #include "process-util.h"
 #include "string-util.h"
+#include "varlink.h"
 #include "vmspawn-register.h"
 
-int register_machine(sd_bus *bus, const char *machine_name, sd_id128_t uuid, const char *service, const char *directory) {
-        _cleanup_(sd_bus_error_free) sd_bus_error error = SD_BUS_ERROR_NULL;
+int register_machine(
+                const char *machine_name,
+                sd_id128_t uuid,
+                const char *service,
+                const char *directory,
+                const char *address,
+                const char *key_path) {
+
+        _cleanup_(varlink_unrefp) Varlink *vl = NULL;
         int r;
 
-        assert(bus);
         assert(machine_name);
         assert(service);
 
-        r = bus_call_method(
-                        bus,
-                        bus_machine_mgr,
-                        "RegisterMachine",
-                        &error,
-                        NULL,
-                        "sayssus",
-                        machine_name,
-                        SD_BUS_MESSAGE_APPEND_ID128(uuid),
-                        service,
-                        "vm",
-                        (uint32_t) getpid_cached(),
-                        strempty(directory));
+        r = varlink_connect_address(&vl, "/run/systemd/machine/io.systemd.Machine");
         if (r < 0)
-                return log_error_errno(r, "Failed to register machine: %s", bus_error_message(&error, r));
+                return log_error_errno(r, "Failed to connect to machined on /run/systemd/machine/io.systemd.Machine: %m");
 
-        return 0;
+        return varlink_callb_and_log(vl,
+                        "io.systemd.Machine.Register",
+                        NULL,
+                        JSON_BUILD_OBJECT(
+                                        JSON_BUILD_PAIR_STRING("name", machine_name),
+                                        JSON_BUILD_PAIR_ID128("id", uuid),
+                                        JSON_BUILD_PAIR_STRING("service", service),
+                                        JSON_BUILD_PAIR_STRING("class", "vm"),
+                                        JSON_BUILD_PAIR_UNSIGNED_NON_ZERO("leader", getpid_cached()),
```

Comment:
if it's just our PID, then i'd not send it along, and instead let the server side derive it automatically from the connection. benefit: we'll use pidfd for this.

(i mean eventually we should come up with a scheme for this how we can literally pass a pidfd, but let's do that some time later.)